--- a/src/build-system/cmake/CMake.NCBIptb.grpc.cmake
+++ b/src/build-system/cmake/CMake.NCBIptb.grpc.cmake
@@ -197,7 +197,9 @@ function(NCBI_internal_process_proto_dataspec _variable _access _value)
             endif()
         endif()
     endif()
-    file(APPEND ${NCBI_GENERATESRC_GRPC} ${_generatesrc})
+    if(NOT NCBI_PTBCFG_PACKAGED AND NOT NCBI_PTBCFG_PACKAGING)
+        file(APPEND ${NCBI_GENERATESRC_GRPC} "${_generatesrc}")
+    endif()
     if (NOT "${_pb_srcfiles}" STREQUAL "")
         NCBI_util_gitignore(${_pb_srcfiles} ${_gr_srcfiles} ${_codeversion})
     endif()
